rust/src/sdk/api.rs

```diff
@@ -0,0 +1,448 @@
+/// one of the following events happen:
+/// - The [`PrimitiveModel`] memory is freed via a call to [`drop_model()`].
+/// - The [`Client`] memory is freed via a call to [`drop_client()`].
+/// - The model is updated via a call to `update_model_N()`.
+/// - The round ends. (TODO: implement this point when a new round is observed)
+pub struct Client {
+    participant: Participant,
+
+    // counting starts from 1, 0 means not seen yet
+    current_round: u32,
+    checked_round: u32,
+
+    // cached primitive model
+    model: Option<PrimitiveModel>,
+}
+
+#[no_mangle]
+/// Creates a new [`Client`].
+pub extern "C" fn new_client() -> *mut Client {
+    let client = Client {
+        participant: if let Ok(participant) = Participant::new() {
+            participant
+        } else {
+            // TODO: add error handling
+            panic!("participant initialization failed")
+        },
+        current_round: 0,
+        checked_round: 0,
+        model: None,
+    };
+    Box::into_raw(Box::new(client))
+}
+
+#[no_mangle]
+/// Destroys a [`Client`] and frees its allocated memory.
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+pub unsafe extern "C" fn drop_client(client: *mut Client) {
+    if !client.is_null() {
+        Box::from_raw(client);
+    }
+}
+
+#[no_mangle]
+/// Checks if the next round has started.
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+pub unsafe extern "C" fn is_next_round(client: *mut Client) -> bool {
+    if client.is_null() {
+        // TODO: add error handling
+        panic!("invalid client");
+    }
+    let client = &mut *client;
+    // TODO: increment the `current_round` if the client sees a new coordinator pk
+    // as a result of `client.handle.get_round_parameters().await`
+    match client.checked_round.cmp(&client.current_round) {
+        // new round since the last check
+        Ordering::Less => {
+            client.checked_round = client.current_round;
+            true
+        }
+        // still same round since the last check
+        Ordering::Equal => false,
+        // should only ever happen if the client is reused for another FL use case
+        Ordering::Greater => panic!("restart the participant for a new FL use case"),
+    }
+}
+
+#[no_mangle]
+/// Checks if the current role of the participant is [`Task::Update`].
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+pub unsafe extern "C" fn is_update_participant(client: *mut Client) -> bool {
+    if client.is_null() {
+        // TODO: add error handling
+        panic!("invalid client");
+    }
+    let client = &mut *client;
```

Comment:
We don't need mutability in this function.

Reply:
good catch, i'll change it. i'm wondering why the compiler didn't say anything.